saml: Don't check existence of templated role names

camscale opened the pull request:
Do not check that a role exists when validating a SAML connector if that
role name contains an expansion - that is, it is a template name using regular
expression capture groups matching against SAML assertion values.

Issue: https://github.com/gravitational/teleport/issues/18593

camscale committed:
saml: Don't check existence of templated role names

Do not check that a role exists when validating a SAML connector if that
role name contains a `$` - that is, it is a template name using regular
expression capture groups matching against SAML assertion values.

Issue: https://github.com/gravitational/teleport/issues/18593

marcoandredinis commented on lib/services/saml.go:
```diff
@@ -103,6 +103,10 @@ func ValidateSAMLConnector(sc types.SAMLConnector, rg RoleGetter) error {
 		for _, mapping := range sc.GetAttributesToRoles() {
 			for _, role := range mapping.Roles {
+				if strings.ContainsRune(role, '$') {
+					// Role is a template so we cannot check for existence of that literal name.
```
Do you think a log (at debug level) would be helpful here?

camscale replied:
I think it might just produce more log spam than utility. Every time a SAML connector is validated (which happens a bit more often than necessary), if that connector uses a template name for a role, we'll log. Not sure that is going to help much with anything, although happy to be convinced otherwise.

marcoandredinis commented on lib/services/saml_test.go:
```diff
@@ -66,3 +69,69 @@ func TestCheckSAMLEntityDescriptor(t *testing.T) {
+	testCases := []struct {
+		desc      string
+		roles     []string
+		expectErr bool
```
:nail_care: 
We could change this to be an `require.ErrorAssertionFunc` :

```suggestion
		errorCheck require.ErrorAssertionFunc
```
Each test would than need a `require.NoError` or 
```go
errorCheck: func(t require.TestingT, err error, i ...interface{}) {
				require.Error(t, err)
				require.ErrorIs(t, <some error>, err)
			},
```

And then, when checking for the error:
```go
tt.errorCheck(t, err)
```

camscale replied:
I'm on the fence about this - I prefer tests to be simple and obvious and I find that using higher-order functions here obscures the test. It's a pretty simple and obvious check as-is and I"m not sure how using the higher-order function here helps. It also means the zero value for the errorCheck function is not useful and it is a value that needs to be supplied to all test cases for the common case.

However I am not actually checking the error type in my test, so I've simplified the tests a little by just using `require.ErrorIs` which works when comparing nil as well as specific errors, removing the conditional and using the zero value effectively.

marcoandredinis replied:
Looks good :+1:

marcoandredinis commented on lib/services/saml.go:
```diff
@@ -103,6 +103,10 @@ func ValidateSAMLConnector(sc types.SAMLConnector, rg RoleGetter) error {
 	if rg != nil {
 		for _, mapping := range sc.GetAttributesToRoles() {
 			for _, role := range mapping.Roles {
+				if strings.ContainsRune(role, '$') {
```
In the linked issue you said that we could replace `$$` with `$` and assume this was a valid role name (and not templating).
It was not clear that we should do that.

Just asking to ensure we didn't forget that.

camscale replied:
I chose not to do that in the end - it started assuming too much about the implementation and I wondered if there should be other substitutions I was not aware of. However I think it becomes moot when I implement r0mant's feedback to use `utils.ContainsExpansion` - the decision is now delegated.

camscale replied:
Actually I'm wrong on that - if I use `utils.ContainsExpansion`, it will return false if the name contains `$$`, so we will try to check if the role name with `$$` is present and will fail. I will have to do the replacement in this case and check the role name.

r0mant commented on lib/services/saml.go:
```diff
@@ -103,6 +103,10 @@ func ValidateSAMLConnector(sc types.SAMLConnector, rg RoleGetter) error {
 	if rg != nil {
 		for _, mapping := range sc.GetAttributesToRoles() {
 			for _, role := range mapping.Roles {
+				if strings.ContainsRune(role, '$') {
```
Let's use [utils.ContainsExpansion](https://github.com/gravitational/teleport/blob/master/lib/utils/replace.go#L26) here instead.

camscale replied:
ack. done

camscale committed:
pr: Use utils.ContainsExpansion, handle $$, simplify tests

Address PR comments, using `utils.ContainsExpansion` instead of
`strings.ContainsRune`. This means we need to handle `$$` as `$` since
that `$$` is false for `ContainsExpansion` but we cannot check that
literal role name without replacing `$$` with `$`.

Simplify the tests and check that the specific error is being returned
by taking advantage of `errors.Is(nil, nil)` returning `true`.

camscale commented:
@r0mant PTAL

camscale committed:
pr: Fix order of parameters to require.ErrorIs

I got them backwards because it is not the usual [expected, actual], but
the other way around.

r0mant commented on lib/services/saml.go:
```diff
@@ -103,6 +103,11 @@ func ValidateSAMLConnector(sc types.SAMLConnector, rg RoleGetter) error {
+					// Role is a template so we cannot check for existence of that literal name.
+					continue
+				}
+				role = strings.ReplaceAll(role, "$$", "$")
```
Let's please add a comment here explaining why this is needed, and also verify this behavior on a real setup with real roles.

camscale replied:
Turns out not to be needed as a role name cannot have a `$` in it, so I've removed this replacement.

camscale committed:
pr; Don't replace $$ with $

A `$` is not valid in a role name, so we do not need to replace `$$` in
an attribute role mapping when checking if the role exists:

    $ tctl create r/role-test-apps.yaml
    ERROR: special characters are not allowed in resource names, please use name composed only from characters, hyphens, dots, and plus signs: "/roles/test$apps/params"

camscale committed:
pr: Remove tests fro `$$` substitution

The substitution was removed, so remove the tests too.

camscale commented:
Note: Not backporting to branch/v9 as originally planned as https://github.com/gravitational/teleport/pull/16565 was not actually backported to v9, despite the label.
